src/GitHub_Updater/Rest_Update.php

```diff
@@ -240,42 +253,70 @@ private function get_bitbucket_webhook_data() {
 		$request_body = file_get_contents('php://input');
 		$request_data = json_decode($request_body, TRUE);
 
-		if (!$request_data) {
-			return NULL;
-		}
-
-		if (!isset($request_data["push"]) ||
-			!isset($request_data["push"]["changes"])) {
+		if ( !$request_data ) {
 			return NULL;
 		}
 
 		$changes = $request_data["push"]["changes"];
 
-		if (!$changes || !sizeof($changes)) {
-			return NULL;
-		}
-
 		// Just use the first entry, assume that it is the right one.
 		$change = $changes[0];
 		$new = $change["new"];
 
 		// What else could this be? For now, just expect branch.
-		if ($new["type"] != "branch") {
+		if ( $new["type"] != "branch" ) {
 			return NULL;
 		}
 
-		$hash = $new["target"]["hash"];
-		$branch = $new["name"];
+		$res = array();
+		$res["webhook"] = "bitbucket";
+		$res["hash"] = $new["target"]["hash"];
+		$res["branch"] = $new["name"];
 
-		return array(
-			"branch" => $branch,
-			"hash" => $hash
-		);
+		return $res;
+	}
+
+	/**
+	 * Check the headers of the request and parse webhook data accordingly.
+	 * This function returns an array containing the elements:
+	 *
+	 *   branch   - The branch that was pushed to.
+	 *   hash     - The most recent hash.
+	 *   webhook  - The type of webhook, i.e. github or bitbucket.
+	 *
+	 * If the request did not come from a webhook, this function returns NULL.
+     *
+	 * We need to rely on the latest commited hash from the remote repository
+	 * and be explicit when specifying the tag we want to update to. If we
+	 * don't do this there is a chance for a race condition, since the default
+	 * zip file on the repository service might not have been created yet.
+	 */
+	private function get_webhook_data() {
+
+		// GitHub
+		if ( $_SERVER["HTTP_X_GITHUB_EVENT"] == "push" ) {
+			return $this->get_github_webhook_data();
```

Comment:
We should rename this function to `id_webhook()`

Then we could have a single method for parsing and pass `$results_data` to smaller functions for each service.

The flow would look something like.

```php
function process_request()
    $this->get_webhook_data();

        $webhook_source = $this->id_webhook;
        $this->parse_webhook_data( $webhook_id, $request_data );

private function parse_webhook_data( $source, $data ) {
    if ( 'gihub' === $source ) {
        $this->parse_github_webhook( $data );
    }
}

or something like that.

In this way we can further simplify each function and keep them with a single responsibility.

Thoughts?